internal/utils/validation.go

```diff
@@ -8,95 +8,129 @@ SPDX-License-Identifier: Apache-2.0
 package utils
 
 import (
-	"crypto/ecdsa"
-	"crypto/x509"
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
 	"fmt"
-	"sort"
-
 	"github.com/hyperledger-labs/fabric-private-chaincode/internal/protos"
 	"github.com/hyperledger/fabric-chaincode-go/shim"
-	"github.com/hyperledger/fabric-protos-go/ledger/rwset/kvrwset"
+	"github.com/hyperledger/fabric/common/flogging"
 )
 
-func ReplayReadWrites(stub shim.ChaincodeStubInterface, rwset *kvrwset.KVRWSet) (readset [][]byte, writeset [][]byte, err error) {
+// #cgo CFLAGS: -I${SRCDIR}/../../common/crypto
+// #cgo LDFLAGS: -L${SRCDIR}/../../common/crypto/_build -L${SRCDIR}/../../common/logging/_build -Wl,--start-group -lupdo-crypto-adapt -lupdo-crypto -Wl,--end-group -lcrypto -lulogging -lstdc++
+// #include <stdio.h>
+// #include <stdlib.h>
+// #include <stdbool.h>
+// #include <stdint.h>
+// #include "pdo-crypto-c-wrapper.h"
+import "C"
+
+var logger = flogging.MustGetLogger("validate")
+
+func ReplayReadWrites(stub shim.ChaincodeStubInterface, fpcrwset *protos.FPCKVSet) (err error) {
 	//TODO error checking
 
 	// nil rwset => nothing to do
+	if fpcrwset == nil {
+		return nil
+	}
+
+	rwset := fpcrwset.GetRwSet()
 	if rwset == nil {
-		return nil, nil, nil
+		return fmt.Errorf("no rwset found")
 	}
 
 	// normal reads
-	var readKeys []string
-	readsetMap := make(map[string][]byte)
-	for _, r := range rwset.Reads {
-		k := TransformToFPCKey(r.Key)
-		readKeys = append(readKeys, k)
-		v, _ := stub.GetState(k)
-		readsetMap[k] = v
+	if rwset.GetReads() != nil {
+		logger.Debugf("Replaying reads")
+		if fpcrwset.GetReadValueHashes() == nil {
+			return fmt.Errorf("no read value hash associated to reads")
+		}
+		if len(fpcrwset.ReadValueHashes) != len(rwset.Reads) {
+			return fmt.Errorf("%d read value hashes but %d reads", len(fpcrwset.ReadValueHashes), len(rwset.Reads))
+		}
 
+		for i := 0; i < len(rwset.Reads); i++ {
+			k := TransformToFPCKey(rwset.Reads[i].Key)
+			v, err := stub.GetState(k)
+			if err != nil {
+				return fmt.Errorf("error (%s) reading key %s", err, k)
+			}
+
+			logger.Debugf("read key %s value(hex) %s", k, hex.EncodeToString(v))
+
+			// compute value hash
+			h := sha256.New()
```

Comment:
For consistency, shouldn't we use here also the pdo hash function (as used in generation?) You already export it through cgo?

Reply:
yes, that export was actually for testing.
I can address this in a future pr. For now it's safe, since with a different hash function would fail

Reply:
added comment